spras/config/schema.py

```diff
@@ -0,0 +1,166 @@
+"""
+Contains the raw pydantic schema for the configuration file.
+
+Using Pydantic as our backing config parser allows us to declaratively
+type our config, giving us more robust user errors with guarantees
+that parts of the config exist after parsing it through Pydantic.
+
+We declare models using two classes here:
+- `BaseModel` (docs: https://docs.pydantic.dev/latest/concepts/models/)
+- `CaseInsensitiveEnum` (see ./util.py)
+"""
+
+import re
+from typing import Annotated, Optional
+
+from pydantic import AfterValidator, BaseModel, ConfigDict, Field
+
+from spras.config.util import CaseInsensitiveEnum
+
+# Most options here have an `include` property,
+# which is meant to make disabling parts of the configuration easier.
+# When an option does not have a default, it means that it must be set by the user.
+
+class SummaryAnalysis(BaseModel):
+    include: bool
+
+    # We prefer to never allow extra keys, to prevent
+    # any user mistypes.
+    model_config = ConfigDict(extra='forbid')
+
+class CytoscapeAnalysis(BaseModel):
+    include: bool
+
+    model_config = ConfigDict(extra='forbid')
+
+# Note that CaseInsensitiveEnum is not pydantic: pydantic
+# has special support for enums, but we avoid the
+# pydantic-specific "model_config" key here for this reason.
+class MlLinkage(CaseInsensitiveEnum):
+    ward = 'ward'
+    complete = 'complete'
+    average = 'average'
+    single = 'single'
+
+class MlMetric(CaseInsensitiveEnum):
+    euclidean = 'euclidean'
+    manhattan = 'manhattan'
+    cosine = 'cosine'
+
+class MlAnalysis(BaseModel):
+    include: bool
+    aggregate_per_algorithm: bool = False
+    components: int = 2
+    labels: bool = True
+    linkage: MlLinkage = MlLinkage.ward
+    metric: MlMetric = MlMetric.euclidean
+
+    model_config = ConfigDict(extra='forbid')
+
+class EvaluationAnalysis(BaseModel):
+    include: bool
+    aggregate_per_algorithm: bool = False
+
+    model_config = ConfigDict(extra='forbid')
+
+class Analysis(BaseModel):
```

Comment:
There is a couple things that get coupled with Evaluation and ML post processing (the logic/coupling is at the bottom of config.py). Is there a way to have a user set the values to true or false based on the coupling if they have evaluation turned to true?

Reply:
Since Pydantic doesn't have a nice solution for dependent fields, getting that system working is a little weird. I've avoided it both here and in #329, but it might be easier just to redesign the config around using nested properties over interdependent fields to get real JSON schema support at that level.